docs/proposals/config_data.md

```diff
@@ -0,0 +1,315 @@
+2. A new pod using the `ConfigData` via the volume plugin is created
+3. The pod is scheduled onto a node
+4. During the sync loop, the Kubelet creates an instance of the volume plugin and calls its
+   `Setup()` method
+5. The volume plugin retrieves the `ConfigData` resource(s) referenced by the pod and projects
+   the appropriate data into the volume
+6. The `ConfigData` referenced by the pod is updated
+7. During the next iteration of the `syncLoop`, the Kubelet creates an instance of the volume plugin
+   and calls its `Setup()` method
+8. The volume plugin projects the updated data into the volume
+
+It is the consuming pod's responsibility to make use of the updated data once it is made visible.
+
+Because environment variables cannot be updated without restarting a container, configuration data
+consumed in environment variables will not be updated.
+
+### Advantages
+
+* Easy to consume in pods; consumer-agnostic
+* Configuration data is persistent and versioned
+* Consumers of configuration data in volumes can respond to changes in the data
+
+## Proposed Design
+
+### API Resource
+
+The `ConfigData` resource will be added to the `extensions` API Group:
+
+```go
+package api
+
+// ConfigData holds configuration data for pods to consume.
+type ConfigData struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+  // Data contains the configuration data.  Each key must be a valid DNS_SUBDOMAIN or leading
+  // dot followed by valid DNS_SUBDOMAIN.
+	Data map[string]string `json:"data,omitempty"`
+}
+
+type ConfigDataList struct {
+	TypeMeta `json:",inline"`
+	ListMeta `json:"metadata,omitempty"`
+
+	Items []ConfigData `json:"items"`
+}
+```
+
+A `Registry` implementation for `ConfigData` will be added to `pkg/registry/configdata`.
+
+### Environment Variables
+
+The `EnvVarSource` will be extended with a new selector for config data:
+
+```go
+package api
+
+// EnvVarSource represents a source for the value of an EnvVar.
+type EnvVarSource struct {
+  // other fields omitted
+
+  // Specifies a ConfigData key
+  ConfigData *ConfigDataSelector `json:"configData,omitempty"`
+}
+
+// ConfigDataSelector selects a key of a ConfigData.
+type ConfigDataSelector struct {
+  // The name of the ConfigData to select a key from.
+  ConfigDataName string `json:"configDataName"`
+  // The key of the ConfigData to select.
+  Key string `json:"key"`
+}
+```
+
+### Volume Source
+
+A new `ConfigDataVolumeSource` type of volume source containing the `ConfigData` object will be
+added to the `VolumeSource` struct in the API:
+
+```go
+package api
+
+type VolumeSource struct {
+  // other fields omitted
+  ConfigData *ConfigDataVolumeSource `json:"configData,omitempty"`
+}
+
+// ConfigDataVolumeSource represents a volume that holds configuration data
+type ConfigDataVolumeSource struct {
+  // A list of config data keys to project into the volume in files
+  Files []ConfigDataVolumeFile `json:"files"`
```

Comment:
I think this is likely to be insufficient.  I think that template expansion is what people really want.  How do we do that?

Do we read a template file from the user's container?  Do we use one configData[key] as the template and the rest as values?  Do we need to support multiple configDatas?

I think we also need to define the final layout in the volume, including how we will handle atomic flips and whether there is a sentinal file that apps can inotify on.

It would be nice to be able to volume mount an expanded template onto a file, but (as we experienced with /etc/hosts) there's not any way to atomically flip that, so I rescind that request, I think.